contentacms/contenta_jsonapi
# Create Release Artifacts on Github on successful release branch builds

Task: #91 


### Description

**TLDR:** Travis CI creates and release successful artifacts to Github on the projects release branch.

Example of releases [here](https://github.com/contentacms/contenta_jsonapi/releases)

##### This Pull Request makes changes to the Travis Build Flow. 

#### Why changes in the flow?
Before, every build had a matrix of jobs that were ran in the same level in a parallel isolated fashion. This flow presented the problem that if one specific job from the matrix was designed to promote the artifact, that job didn't have the knowledge of the other jobs status, so artifacts could be promoted with failed build, and we want
to prevent that.

#### What is the change?
 Now, every build have stages, stages are an isolated group of jobs. We have a Testing Stage, which runs all the test and a Release Stage which promotes the artifact. The advantage of this flow is that the Stages are run sequentially, are dependent of each other success and their status depends on all the jobs inside of it. If one job fails on the Test Stage, the Release Stage won't be run, though no artifact will be promoted and only artifacts that all tests passed are released.

#### What is needed?
There are a few environment variables that need to be setup in the Travis Backend, they are already configured, which makes this proccess flexible.
This Environment variables are:

- **SECURE_TOKEN**: A github, non travis encrypted, token is needed for uploading the artifact to github.
- **PROJECT_RELEASE_BRANCH**: The name of the designated branch that artifacts are going to be based on. This branch should be consider as the release branch for the project, this branch should exist on the Travis branch list. **Any changes on this branch will trigger an artifact release**.

### Notes
Currently we are treating the branch `8.x-1.x` as our release branch, unfortunately Travis runs on branches specified in the branch stanza, [here](https://github.com/contentacms/contenta_jsonapi/blob/8.x-1.x/.travis.yml#L12) for example. For this reason, `PROJECT_RELEASE_BRANCH` is defaulted to `8.x-1.x`, in a future when more branches are added, and the release branch gets created, the `PROJECT_RELEASE_BRANCH` just needs to change in the Travis Bankend Environment Variable.

---

An initial attempt of creating Github issues on failed build was done. Currently Travis doesn't support this feature as stated [here](https://github.com/travis-ci/travis-ci/issues/7199#issuecomment-313101882), it is still a future feature request. The reason why is a bit complex is due to that we have a matrix of isolated jobs, this can cause the creation of multiple github issues, one for each job failed in the matrix, for one build, for this reason it can't be job specific. Researched on trying in it in a Stage specific matter for the new flow, but Travis doesn't have Stage events or hooks. The conclusion is based on the current status and Travis limitations, we can only create issues, in the new flow, that reflect that the artifact got released or not. Since Travis Stages run sequentially and only if the previous Stage was successful, if a job fails, the next Stage doesn't get trigger, it complicates the situation.
For this feature if desired, we have to live with one of the following:
- Multiple issues for one build
- Issue if a good artifact doesn't get released.
- No issues get released, abandon the feature.

***Note**: this feature is currently turned off, for turning it on set `GITHUB_CREATE_ISSUE_FEATURE` environment variable to true in the Travis Backend.
 ***_If desired it can be moved to another branch and open an new enhancement issue._**

### Test Instructions

Many different way of testing this, the following is the preferred one.
Create 

1. another branch brased on `feature/91_Generate_GitHub_Releases`.
2. Change the `PROJECT_RELEASE_BRANCH` on the Travis CI Environment Variable [here](https://travis-ci.org/contentacms/contenta_jsonapi/settings)
3. Commit and Push a simple change in a file, a new comment for example.
4. Ensure that Travis runs.
5. When the Build finishes, verify that a new artifact was released [here](https://github.com/contentacms/contenta_jsonapi/releases)
 6. Delete the artifact and the tag.
